File: dp/451-sort-characters-by-frequency/sort-characters-by-frequency.go
type pair struct {
	char  byte
	count int
}

func frequencySort(s string) string {
	count := make(map[byte]int)

	// Count the frequency of each character
	for i := 0; i < len(s); i++ {
		count[s[i]]++
	}

	// Create a slice of pairs from the map
	var pairs []pair
	for char, freq := range count {
		pairs = append(pairs, pair{char, freq})
	}

	// Sort the pairs based on their count in descending order
	sort.Slice(pairs, func(i, j int) bool {
		return pairs[i].count > pairs[j].count
	})

	// Construct the sorted string
	var sortedString []byte
	for _, p := range pairs {
		for i := 0; i < p.count; i++ {
			sortedString = append(sortedString, p.char)
		}
	}

	return string(sortedString)
}

File: dp/451-sort-characters-by-frequency/sort-characters-by-frequency.py
class Solution:
    def frequencySort(self, s: str) -> str:
        count = {}
   
        for n in s:
            count[n] = 1 + count.get(n, 0)

        print(count) 
        a = sorted(count.items(), key=lambda x: x[1], reverse=True)    
        print(a)
        st = ""
        for i in a:
            print(i[0])
            st+=i[0]*i[1]

        print(st)

        return st